### PR TITLE
feat(KNO-4578) Add support for knock commit promote --only flag

### PR DESCRIPTION
**Pull request by francoborr**
### Description
This PR adds a `--only` flag to the `knock commit promote` command.

With the addition of this PR, the command will start having two modes of operation:
* If `--to` is provided, it will promote all changes to the destination environment
* If `--only` is provided, it will promote the given commit based on the commit ID to the subsequent environment.
* If both flags are present, we return an error. We currently don't support promoting one commit into a given destination.
* One of both flags must be present, if none of them is set then we also return an error.

### Tasks
[KN0-4578](https://linear.app/knock/issue/KNO-4578/[cli]-add-support-for-knock-commit-promote-only-command)

### CODE REVIEW

File: src/commands/commit/promote.ts [hunk 1]
```diff
@@ -6,34 +6,66 @@ import { withSpinner } from "@/lib/helpers/request";
...
-    // Confirm first as we are about to promote changes to go live in the target
-    // environment, unless forced.
-    const prompt = `Promote all changes to \`${flags.to}\` environment?`;
-    const input = flags.force || (await promptToConfirm(prompt));
-    if (!input) return;
+    // Currently this command support two modes of operation based on the flags:
+    // * --to : Promotes all changes to the destination environment.
+    // * --only: Promotes one commit to the subsequent enviroment
+    // The absence or presence of both flags will result in an error.
+
```
Comment (francoborr) [inline src/commands/commit/promote.ts:32]: I've open this PR as a brief so we can confirm this is the expected behavior:

This command has now the two operation modes based on the flags:
* --to : Promotes all changes to the destination environment.
 * --only: Promotes one commit to the subsequent enviroment
This also means that the `--to` and `--only` flags can't be used at the same time, but at least one of them must be set.

The solution I found is to check the presence/absence of these here, and based on the flag received call `promoteAllChanges` or `promoteCommitChanges `

The good thing of this approach is that we can encapsulate both options (promoting one commit / promoting all changes) in one single command, but the downside is that we need to do this custom handle here.

Another option could be to separate this into two commands:
* knock commit promote-all (or only promote)
* knock commit promote-one

File: test/commands/commit/promote.test.ts [hunk 2]
```diff
@@ -46,4 +76,26 @@ describe("commands/commit/promote", () => {
...
+      .it("calls apiV1 promoteChange with expected props", () => {
+        sinon.assert.calledWith(
+          KnockApiV1.prototype.promoteChange as any,
+          sinon.match(({ flags }) =>
+            isEqual(flags, {
+              "service-token": "valid-token",
+              only: "example-id",
+            }),
+          ),
+          sinon.match((commit) =>
```
Comment (francoborr) [inline test/commands/commit/promote.test.ts:93]: I don't know why this test fails, I checked and the `commit` is returned when calling `promoteChange` 🧐

File: src/commands/commit/promote.ts [hunk 3]
```diff
@@ -2,38 +2,82 @@ import { Flags } from "@oclif/core";
...
 
-    this.log(
-      `‣ Successfully promoted all changes to \`${flags.to}\` environment`,
-    );
+      const { commit } = resp.data;
+      this.log(
+        `‣ Successfully promoted commit \`${flags.only}\` into \`${commit?.environment}\` environment`,
+      );
+      this.log(`‣ New commit ID: \`${commit?.id}\``);
+    }
```
Comment (thomaswhyyou) [inline src/commands/commit/promote.ts:81]: This is a small formatting thing but can we match this message format to how we are doing for the workflow run command? https://github.com/knocklabs/knock-cli/blob/6ff133cd9723ebd296dc34d2a83762f935a11fc5/src/commands/workflow/run.ts

So something like:
```
this.log(
  `‣ Successfully promoted the commit \`${flags.only}\` into \`${commit!.environment}\` environment`,
);
this.log(indentString(`Commit id: ${commit!.id}`, 4));
```

**Review by thomaswhyyou — COMMENTED** [review]
A few small things, but overall makes sense to me.

File: src/lib/api-v1.ts [hunk 4]
```diff
@@ -160,6 +160,12 @@ export default class ApiV1 {
     return this.put(`/commits/promote`, {}, { params });
   }
 
+  async promoteChange({
+    flags,
+  }: Props): Promise<AxiosResponse<PromoteChangeResp>> {
+    return this.put(`/commits/${flags.only}/promote`, {}, {});
```
[the remark this answers is not in the document]
Comment (francoborr) [reply src/lib/api-v1.ts:166]: Is not, thanks for the nit!

File: src/commands/commit/promote.ts [hunk 5]
```diff
@@ -2,38 +2,82 @@ import { Flags } from "@oclif/core";
...
+        const message = formatErrorRespMessage(resp);
+        return this.error(new ApiError(message));
+      }
 
-    this.log(
-      `‣ Successfully promoted all changes to \`${flags.to}\` environment`,
-    );
+      const { commit } = resp.data;
+      this.log(
+        `‣ Successfully promoted commit \`${flags.only}\` into \`${commit?.environment}\` environment`,
```
[the remark this answers is not in the document]
Comment (francoborr) [reply src/commands/commit/promote.ts:78]: Yep that's right, if not it means there was an error (in which case we will never get this far).

[the remark this answers is not in the document]
Comment (thomaswhyyou) [reply src/commands/commit/promote.ts:78]: Then no reason have the `?` in `commit?`.

[the remark this answers is not in the document]
Comment (francoborr) [reply src/commands/commit/promote.ts]: Yes you are right, thanks!

File: src/commands/commit/promote.ts [hunk 6]
```diff
@@ -1,27 +1,82 @@
...
+  isSuccessResp,
+  withSpinner,
+} from "@/lib/helpers/request";
+import { indentString } from "@/lib/helpers/string";
 import { promptToConfirm } from "@/lib/helpers/ux";
 
 export default class CommitPromote extends BaseCommand<typeof CommitPromote> {
-  static summary = "Promote all changes to the destination environment.";
+  static summary =
+    "Promote a single commit to the subsequent environment or all changes to the destination environment.";
```
Comment (thomaswhyyou) [inline src/commands/commit/promote.ts:16]:
```suggestion
    "Promote one or all commits to the subsequent environment.";
```

File: src/commands/commit/promote.ts [hunk 7]
```diff
@@ -1,27 +1,82 @@
...
       summary:
-        "The destination environment to promote changes from the preceding environment.",
-      required: true,
+        "The destination environment to promote all changes from the preceding environment.",
     }),
     force: Flags.boolean({
       summary: "Remove the confirmation prompt.",
     }),
+    only: Flags.string({
+      summary: "The ID of the commit to promote to the subsequent environment",
```
Comment (thomaswhyyou) [inline src/commands/commit/promote.ts:27]:
```suggestion
      summary: "The target commit id to promote to the subsequent environment",
```

File: src/commands/commit/promote.ts [hunk 8]
```diff
@@ -1,27 +1,82 @@
...
+
+    if (!flags.to && !flags.only) {
+      throw new Error(
+        "You must specify either `--to` or `--only` flag.\n See more help with --help",
+      );
+    }
+
+    return flags.only
+      ? this.promoteOneChange(flags)
+      : this.promoteAllChanges(flags);
```
Comment (thomaswhyyou) [inline src/commands/commit/promote.ts:52]:
```suggestion
      ? this.promoteOneChange()
      : this.promoteAllChanges();
```
Nit, you don't need to pass in flags. You can access them via `this.props.flags`.

File: src/lib/base-command.ts [hunk 9]
```diff
@@ -10,7 +10,9 @@ export type BFlags = Interfaces.InferredFlags<
   (typeof BaseCommand)["baseFlags"]
 >;
 
-type TFlags<T extends typeof Command> = Interfaces.InferredFlags<T["flags"]> &
+export type TFlags<T extends typeof Command> = Interfaces.InferredFlags<
```
Comment (thomaswhyyou) [inline src/lib/base-command.ts:13]: Probably unncessary to export with the change I suggested above.

**Review by thomaswhyyou — APPROVED** [review]
Awesome! A couple small copy edits and one follow up suggestion, but this looks good to go 👍

File: src/commands/commit/promote.ts (hunk 8 above)
Comment (francoborr) [reply src/commands/commit/promote.ts:52]: Gotcha!

**Pull request merged**